### ISSUE TITLE
FileWatch with std::filesystem::path does not compile

### ISSUE THREAD

**Issue by SC5Shout** [ThomasMonkman/filewatch#39]
OS: Windows
Compiler: MSVC - the latest one, don't know the number

```cpp
#include <filesystem>
#include <FileWatch.hpp>
#include <iostream>
#include <string>
#include <iostream>

int main()
{
	filewatch::FileWatch<std::filesystem::path> watch(
		"./",
		[](const std::filesystem::path& path, const filewatch::Event change_type) {
		std::cout << path << "\n";
	}
	);

	while (true) {

	}
}
```
Severity	Code	Description	Project	File	Line	Suppression State
Error	C2440	'<function-style-cast>': cannot convert from 'initializer list' to 'StringType'	D:\Projects\Test\filewatch\FileWatch.hpp	741	

however this code works:

```cpp
#include <filesystem>
#include <FileWatch.hpp>
#include <iostream>
#include <string>
#include <iostream>

int main()
{
	filewatch::FileWatch<std::string> watch(
		"./",
		[](const std::filesystem::path& path, const filewatch::Event change_type) {
			std::cout << path << "\n";
		}
	);

	while (true) {

	}
}
```

**Comment by ThomasMonkman** [comment]
Ok thank you, I wrote this back when std::filesystem support was flaky, I will have a look

**Comment by Lumengine** [comment]
Hi and thank you for this helper !  I have the same issue here !

**Comment by richardeakin** [comment]
I'm seeing the same error when attempting the build using the laster master FileWatch.hpp + MSVC from Visual Studio 2022 latest, specifically building using v143 toolset and either `/std:c++17` or `/std:c++20` language standards.

Looking at `std::filesystem::path`'s [constructors](https://en.cppreference.com/w/cpp/filesystem/path/path), I don't see one that can take a char* buffer + length. Think it needs to be converted to a string first but I didn't see how you define which string type to use when `std::filesystem::path` is being used (it does appear to define a `string_type` that could be used).
 
Full compile error output:
```shell
30>G:\code\graphics\diligent\diligent-playground\test\standalone\Terrain\src\../../common/src/FileWatch.hpp(740,30): error C2440: '<function-style-cast>': cannot convert from 'initializer list' to 'StringType'
30>        with
30>        [
30>            StringType=std::filesystem::path
30>        ] (compiling source file G:\code\graphics\diligent\diligent-playground\test\standalone\Terrain\src\Terrain.cpp)
30>G:\code\graphics\diligent\diligent-playground\test\standalone\Terrain\src\../../common/src/FileWatch.hpp(740,30): message : No constructor could take the source type, or constructor overload resolution was ambiguous (compiling source file G:\code\graphics\diligent\diligent-playground\test\standalone\Terrain\src\Terrain.cpp)
30>G:\code\graphics\diligent\diligent-playground\test\standalone\Terrain\src\../../common/src/FileWatch.hpp(727,1): message : while compiling class template member function 'StringType filewatch::FileWatch<StringType>::absolute_path_of(const StringType &)'
30>        with
30>        [
30>            StringType=std::filesystem::path
30>        ] (compiling source file G:\code\graphics\diligent\diligent-playground\test\standalone\Terrain\src\Terrain.cpp)
30>G:\code\graphics\diligent\diligent-playground\test\standalone\Terrain\src\../../common/src/FileWatch.hpp(182,18): message : see reference to function template instantiation 'StringType filewatch::FileWatch<StringType>::absolute_path_of(const StringType &)' being compiled
30>        with
30>        [
30>            StringType=std::filesystem::path
30>        ] (compiling source file G:\code\graphics\diligent\diligent-playground\test\standalone\Terrain\src\Terrain.cpp)
30>C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.35.32215\include\memory(3065,1): message : see reference to class template instantiation 'filewatch::FileWatch<std::filesystem::path>' being compiled (compiling source file G:\code\graphics\diligent\diligent-playground\test\standalone\Terrain\src\Terrain.cpp)
30>C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.35.32215\include\memory(3064,1): message : while compiling class template member function 'void std::default_delete<filewatch::FileWatch<std::filesystem::path>>::operator ()(_Ty *) noexcept const'
30>        with
30>        [
30>            _Ty=filewatch::FileWatch<std::filesystem::path>
30>        ] (compiling source file G:\code\graphics\diligent\diligent-playground\test\standalone\Terrain\src\Terrain.cpp)
30>C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.35.32215\include\memory(3177,33): message : see reference to function template instantiation 'void std::default_delete<filewatch::FileWatch<std::filesystem::path>>::operator ()(_Ty *) noexcept const' being compiled
30>        with
30>        [
30>            _Ty=filewatch::FileWatch<std::filesystem::path>
30>        ] (compiling source file G:\code\graphics\diligent\diligent-playground\test\standalone\Terrain\src\Terrain.cpp)
30>C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.35.32215\include\memory(3101): message : see reference to class template instantiation 'std::default_delete<filewatch::FileWatch<std::filesystem::path>>' being compiled (compiling source file G:\code\graphics\diligent\diligent-playground\test\standalone\Terrain\src\Terrain.cpp)
30>G:\code\graphics\diligent\diligent-playground\test\standalone\Terrain\src\Terrain.cpp(63,62): message : see reference to class template instantiation 'std::unique_ptr<filewatch::FileWatch<std::filesystem::path>,std::default_delete<filewatch::FileWatch<std::filesystem::path>>>' being compiled
```

**Comment by jdunlap** [comment]
It compiles and works fine for me IF you pass an actual `std::filesystem::path` as the first parameter to the constructor instead of a string.

So, change your code to:

```c++
int main()
{
	filewatch::FileWatch<std::string> watch(
		std::filesystem::path("./"),
		[](const std::filesystem::path& path, const filewatch::Event change_type) {
			std::cout << path << "\n";
		}
	);

	while (true) {

	}
}
```

and it should work.

**Comment by richardeakin** [comment]
Yes it does work when I test now (with C++20 too), as long as you pass in a `std::filesystem::path` explicitly. Thanks for pointing that out.

**Comment by RichardLuo0** [comment]
With clang 18.1.8 on Windows, the `filewatch::FileWatch<std::filesystem::path>(std::filesystem::path("./"), [](const std::filesystem::path&, const filewatch::Event) {})` will not compile. With `filewatch::FileWatch<std::wstring>(std::filesystem::path("./"), [](const std::filesystem::path&, const filewatch::Event) {})` the program will crash inside `absolute_path_of`. However if I change it to std::filesystem::absolute(), it will work but the callback gets called twice for one save.

**Comment by nepp95** [comment]
This seems to still be an issue although I can remember having used it in the past without any problems. Compiling with C++17 on MSVC 14.39

My specific implementation which should be simple enough not to cause trouble:

```
// Defined somewhere else:
// m_FileWatcher has type: filewatch::FileWatch<std::filesystem::path>
// filepath has type: const std::filesystem::path&, since this is passed into the method by reference.

m_FileWatcher = filewatch::FileWatch<std::filesystem::path>(filepath, OnFileSystemEvent);

// Method referred in constructor
static void OnFileSystemEvent(const std::filesystem::path& filepath, const filewatch::Event changeType)
{
    // Do stuff
}
```

Yet exactly the same error: '<function style cast>': cannot convert from 'initializer list' to 'StringType', line 741.

Having looked back at earlier projects where this did just work and comparing the `FileWatch.hpp`, I can see that the filewatcher has a templated argument for the first argument, yet on this argument it now tries to call `absolute_path_of` in the initializer list, where it didn't before. When using `std::filesystem::path` this argument gets converted to a wchar_t when using MSVC but the method `absolute_path_of` then tries to call `.c_str()` on a wchar_t, which is not possible.

Since I'm not too deep into the internals, I ofcourse can't really advocate for the choices being made or I might just be completely missing something. All is possible, but this is as far as I have traced back the problem. I can also be misreading this since I am not so much into templated arguments myself usually.

Where this did work was in this commit: https://github.com/ThomasMonkman/filewatch/commit/9237db34257de1de1bbeddecae1c2250d2e16cb0

The pull request after this seems to have broken this. I have also confirmed my exact code from above works fine with the version of that commit.
